flycontrols.js

```diff
@@ -0,0 +1,183 @@
+  this.type = 'Flycontrols'
+  this.enabled = true
+  this.changed = new Signal()
+  this.entity = null
+  this.dirty = false
+  this.camera =  false
+  this.currentYaw = [0,0,0,1]
+  this.currentPitch = [0,0,0,1]
+
+  this.isLocked = false
+
+  this.moveForward = false 
+  this.moveLeft = false
+  this.moveBackward = false
+  this.moveRight = false
+  this.moveUp = false
+  this.moveDown = false
+
+  const initialState = {
+    position: [0, 1, 0],
+    element: opts.element || document,
+    flySpeed : 0.1,
+    sensitivityX : 0.005,
+    sensitivityY : 0.005,
+    developerMode : false
+  }
+
+  this.set(initialState)
+  this.set(opts)
+}
+
+Flycontrols.prototype.init = function (entity) {
+  this.entity = entity
+  this.setup()
+}
+
+Flycontrols.prototype.set = function (opts) {
+  Object.assign(this, opts)
+  Object.keys(opts).forEach((prop) => this.changed.dispatch(prop))
+}
+
+Flycontrols.prototype.update = function () {
+  if(this.moveForward || 
+    this.moveBackward || 
+    this.moveLeft || 
+    this.moveRight ||
+    this.moveUp || 
+    this.moveDown ){
+    let dir = [0,0,0]; 
+
+    if(this.moveForward ) vec3.add(dir,[0,0,-1])
+    if(this.moveBackward) vec3.add(dir,[0,0,1])
+    if(this.moveLeft) vec3.add(dir,[-1,0,0])
+    if(this.moveRight) vec3.add(dir,[1,0,0])
+
+    let vec = vec3.multQuat(dir,this.entity.transform.rotation)
+
+    if(this.moveUp) vec3.add(vec,[0,1,0])
+    if(this.moveDown) vec3.add(vec,[0,-1,0])
+
+    vec3.normalize(vec)
+    vec3.scale(vec,this.flySpeed)
+    let newPos = vec3.add(this.entity.transform.position,vec)
+    this.entity.transform.set({ position : newPos})
+  }
+
+}
+
+Flycontrols.prototype.updateWindowSize = function () {
+
+}
+
+
+Flycontrols.prototype.setup = function () {
+  var flycontrols = this;
+  if(!this.camera) this.camera = this.entity.getComponent('Camera') || false
+  
+  this.entity.transform.set({ position : flycontrols.position})
+
+  function onMouseDown (e) {
+    if (!flycontrols.enabled) return
+    if (flycontrols.developerMode && e.button == 0)document.body.requestPointerLock();
+
+  }
+
+  function onMouseMove (e) {
+    if (!flycontrols.enabled) return
+    if(document.pointerLockElement){
+        var movementX = e.movementX || e.mozMovementX || e.webkitMovementX || 0;
+        var movementY = e.movementY || e.mozMovementY || e.webkitMovementY || 0;
+
+        quat.mult(flycontrols.currentYaw, quat.fromAxisAngle(quat.create(), [0,1,0], -movementX * flycontrols.sensitivityX))
+        quat.mult(flycontrols.currentPitch, quat.fromAxisAngle(quat.create(), [1,0,0], -movementY * flycontrols.sensitivityY))
+
+        let newRot = quat.mult(quat.mult([0,0,0,1],flycontrols.currentYaw),flycontrols.currentPitch)
+        
+        flycontrols.entity.transform.set({rotation : newRot})
+    }
+  }
+
+  function onMouseUp (e) {
+    if (!flycontrols.enabled) return
+    !flycontrols.developerMode ? document.body.requestPointerLock() :  document.exitPointerLock();
```

Comment:
This should be not needed if we requestPointerLock on mouseDown, seem a bit convoluted logic

Reply:
Sorted